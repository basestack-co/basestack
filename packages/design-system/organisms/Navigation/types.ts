export interface ButtonLinkProps {
  /**
   * Set button link active
   */
  isActive: boolean;
}

export interface NavigationProps {
  /**
   * create flag callback
   */
  onCreateFlag: () => void;
}
