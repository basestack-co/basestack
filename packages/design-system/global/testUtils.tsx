import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
