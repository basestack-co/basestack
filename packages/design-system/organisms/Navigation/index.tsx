import React, { memo } from "react";
import { useTheme } from "styled-components";
import { Button, Avatar } from "../../atoms";
import {
  Container,
  List,
  ListItem,
  LogoContainer,
  ButtonContainer,
} from "./styles";

interface ButtonLinkProps {
  isActive: boolean;
}

const ButtonLink: React.FC<ButtonLinkProps> = ({
  isActive = false,
  children,
}) => (
  <ButtonContainer isActive={isActive}>
    <Button as="a" variant="primaryNeutral">
      {children}
    </Button>
  </ButtonContainer>
);

const Navigation = () => {
  const theme = useTheme();

  return (
    <Container data-testid="navigation">
      <List data-testid="navigation-left-ul">
        <ListItem>
          <LogoContainer>
            <Avatar alt="user image" userName="Logo" />
          </LogoContainer>
        </ListItem>
        <ListItem>
          <Button
            as="a"
            iconPlacement="right"
            icon="expand_more"
            variant="primaryNeutral"
          >
            Project Name
          </Button>
        </ListItem>
        <ListItem>
          <ButtonLink isActive>Features</ButtonLink>
        </ListItem>
        <ListItem>
          <ButtonLink isActive={false}>Settings</ButtonLink>
        </ListItem>
        <ListItem ml={theme.spacing.s3}>
          <Button variant="primary">Create flag</Button>
        </ListItem>
      </List>
      <List ml="auto" data-testid="navigation-right-ul">
        <ListItem>
          <ButtonLink isActive={false}>Documentation</ButtonLink>
        </ListItem>
        <ListItem>
          <ButtonLink isActive={false}>Resources</ButtonLink>
        </ListItem>
        <ListItem>
          <Button as="a" variant="primaryNeutral">
            Github
          </Button>
        </ListItem>
        <ListItem ml={theme.spacing.s3}>
          <Avatar alt="user image" userName="FlÃ¡vio Amaral" />
        </ListItem>
      </List>
    </Container>
  );
};

export default memo(Navigation);
