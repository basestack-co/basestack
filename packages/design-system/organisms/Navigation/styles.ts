import styled, { css } from "styled-components";
import { space } from "styled-system";
import { rem } from "polished";

export const Container = styled.nav`
  position: sticky;
  top: 0;
  z-index: ${({ theme }) => theme.zIndex.navigation};
  display: flex;
  align-items: center;
  background-color: ${({ theme }) => theme.colors.white};
  box-shadow: ${({ theme }) => theme.shadow.elevation2};
  height: ${rem("64px")};
  padding: 0 ${({ theme }) => theme.spacing.s4};
`;

const flexCenter = css`
  display: flex;
  align-items: center;
`;

export const ButtonContainer = styled.div<{ isActive: boolean }>`
  ${flexCenter};
  position: relative;
  height: ${rem("64px")};
  ${({ isActive }) =>
    isActive &&
    css`
      a {
        color: ${({ theme }) => theme.colors.blue400};
      }

      &::before {
        content: "";
        position: absolute;
        right: 0;
        bottom: 0;
        left: 0;
        height: 3px;
        background-color: ${({ theme }) => theme.colors.blue400};
      }
    `};
`;

export const List = styled.ul`
  ${space};
  ${flexCenter};
`;

export const ListItem = styled.li`
  ${space};
`;

export const LogoContainer = styled.div`
  ${flexCenter};
  position: relative;
  padding-right: ${({ theme }) => theme.spacing.s4};
  margin-right: ${({ theme }) => theme.spacing.s1};

  &::before {
    content: "";
    position: absolute;
    height: 20px;
    right: -1px;
    width: 2px;
    background-color: ${({ theme }) => theme.colors.gray100};
  }
`;
