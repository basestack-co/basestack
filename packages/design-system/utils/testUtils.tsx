import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
