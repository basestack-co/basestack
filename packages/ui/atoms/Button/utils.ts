import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
