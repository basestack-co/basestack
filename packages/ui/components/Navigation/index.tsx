import React, { Fragment } from "react";
// Types
import {
  DesktopNavigationProps,
  MobileNavigationProps,
  GeneralNavigationProps,
} from "./types";
// Components
import DesktopNavigation from "./Desktop";
import MobileNavigation from "./Mobile";

export interface NavigationProps extends GeneralNavigationProps {
  desktopProps: DesktopNavigationProps;
  mobileProps: MobileNavigationProps;
}

const Navigation = ({
  desktopProps,
  mobileProps,
  ...props
}: NavigationProps) => {
  return (
    <Fragment>
      <DesktopNavigation {...desktopProps} {...props} />
      <MobileNavigation {...mobileProps} {...props} />
    </Fragment>
  );
};

export default Navigation;
