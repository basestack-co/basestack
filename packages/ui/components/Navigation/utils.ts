// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
