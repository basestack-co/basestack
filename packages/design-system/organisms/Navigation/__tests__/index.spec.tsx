import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Navigation from "..";

describe("Navigation Organism tests", () => {
  afterEach(cleanup);

  test("should render Navigation correctly", () => {
    const { asFragment } = renderWithTheme(
      <Navigation onCreateFlag={jest.fn} pathname="/flags" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
