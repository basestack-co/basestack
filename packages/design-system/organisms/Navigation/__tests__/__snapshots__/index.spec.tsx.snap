// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Organism tests should render Navigation correctly 1`] = `
<DocumentFragment>
  .c3 {
  height: 2.5rem;
  width: 2.5rem;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  height: 100%;
  width: 100%;
  object-fit: cover;
}

.c6 {
  height: 1.5rem;
  width: 1.5rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-left: 0.25rem;
}

.c7 {
  color: #000000;
  font-size: 1.5rem;
}

.c5 {
  border: none;
  height: 2.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 0.75rem;
  font-size: 0.875rem;
  font-weight: 500;
  cursor: pointer;
  border-radius: 4px;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: transparent;
  color: #000000;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  padding-right: 0.5rem;
}

.c5 .material-symbols-sharp {
  color: inherit;
}

.c5:hover {
  color: #276EF1;
  background-color: #EFF3FE;
}

.c5:active {
  background-color: #D4E2FC;
}

.c9 {
  border: none;
  height: 2.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 0.75rem;
  font-size: 0.875rem;
  font-weight: 500;
  cursor: pointer;
  border-radius: 4px;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: transparent;
  color: #000000;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c9 .material-symbols-sharp {
  color: inherit;
}

.c9:hover {
  color: #276EF1;
  background-color: #EFF3FE;
}

.c9:active {
  background-color: #D4E2FC;
}

.c12 {
  border: none;
  height: 2.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 0.75rem;
  font-size: 0.875rem;
  font-weight: 500;
  cursor: pointer;
  border-radius: 4px;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: #276EF1;
  color: #FFFFFF;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c12:hover:not(:active) {
  background-color: #0e55d7;
}

.c0 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 100;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  box-shadow: rgba(0,0,0,0.1) 0px 1px 3px 0px,rgba(0,0,0,0.06) 0px 1px 2px 0px;
  height: 4rem;
  padding: 0 1.25rem;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  height: 4rem;
}

.c8 a {
  color: #276EF1;
}

.c8::before {
  content: "";
  position: absolute;
  right: 0;
  bottom: 0;
  left: 0;
  height: 3px;
  background-color: #276EF1;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  height: 4rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c13 {
  margin-left: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  margin-left: 0.75rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  padding-right: 1rem;
  margin-right: 0.25rem;
}

.c2::before {
  content: "";
  position: absolute;
  height: 20px;
  right: -1px;
  width: 2px;
  background-color: #EEEEEE;
}

<nav
    class="c0"
    data-testid="navigation"
  >
    <ul
      class="c1"
      data-testid="navigation-left-ul"
    >
      <li
        class=""
      >
        <div
          class="c2"
        >
          <div
            class="c3"
            data-testid="avatar"
            size="40"
          >
            <img
              alt="user image"
              class="c4"
              src="https://avatars.dicebear.com/api/initials/Logo.svg?size=40&fontSize=40"
            />
          </div>
        </div>
      </li>
      <li
        class=""
      >
        <a
          class="c5"
        >
          Project Name
          <div
            class="c6"
            data-testid="icon-container"
            size="24"
          >
            <span
              class="c7 material-symbols-sharp"
              data-testid="icon"
              size="24"
            >
              expand_more
            </span>
          </div>
        </a>
      </li>
      <li
        class=""
      >
        <div
          class="c8"
        >
          <a
            class="c9"
          >
            Features
          </a>
        </div>
      </li>
      <li
        class=""
      >
        <div
          class="c10"
        >
          <a
            class="c9"
          >
            Settings
          </a>
        </div>
      </li>
      <li
        class="c11"
      >
        <button
          class="c12"
        >
          Create flag
        </button>
      </li>
    </ul>
    <ul
      class="c13"
      data-testid="navigation-right-ul"
    >
      <li
        class=""
      >
        <div
          class="c10"
        >
          <a
            class="c9"
          >
            Documentation
          </a>
        </div>
      </li>
      <li
        class=""
      >
        <div
          class="c10"
        >
          <a
            class="c9"
          >
            Resources
          </a>
        </div>
      </li>
      <li
        class=""
      >
        <a
          class="c9"
        >
          Github
        </a>
      </li>
      <li
        class="c11"
      >
        <div
          class="c3"
          data-testid="avatar"
          size="40"
        >
          <img
            alt="user image"
            class="c4"
            src="https://avatars.dicebear.com/api/initials/FlÃ¡vio Amaral.svg?size=40&fontSize=40"
          />
        </div>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
